app/models/spree/calculator/shipping/store_distance.rb

```diff
@@ -0,0 +1,89 @@
+require 'pry'
+require_dependency 'spree/shipping_calculator'
+
+module Spree
+  module Calculator::Shipping
+    class StoreDistance < Spree::ShippingCalculator
+      preference :base_amount_up_to_2_5_mi, :decimal, default: 2.5
+      preference :base_amount_up_to_3_5_mi, :decimal, default: 3.5
+      preference :base_amount_up_to_6_mi, :decimal, default: 4.5
+      preference :base_amount_up_to_9_mi, :decimal, default: 5.5
+
+      def self.description
+        Spree.t(:shipping_store_distance)
+      end
+
+      def compute_package(package)
+        compute_store_distance(package)
+      end
+
+      def vendors_coordinates(package)
+        package.order.variants.map do |variant|
+          variant.vendor.to_coordinates
+        end
+      end
+
+      def total_product_weight(package)
+        total_weight = 0
+
+        total_weight = package.order.variants.map do |variant|
+          variant.weight
+        end
+        total_weight.sum
+      end
+
+      def total_distance(package)
+        vendor_co = vendors_coordinates(package)
+
+        total_distance = 0
+
+        total_distance += Geocoder::Calculations.distance_between(package.stock_location.to_coordinates, vendor_co[0])
+
+        vendor_co[0...-1].each_with_index do |val, index|
```

Comment:
https://til.hashrocket.com/posts/52tcyljcgl-reducing-with-an-index-in-ruby

I think you can extend vendor_coordinates array in a way you won't have to calculate, total_distance, outside this loop.